components/st-microelectronics/STM32F407ZET6.stanza

```diff
@@ -0,0 +1,534 @@
+    [ VSS[5]         |  94 | Down  | supply | -             | -             | "VSS"            ] ;| S   | -   | -      | -                                                                                                                   | -
+    [ VSS[6]         |  61 | Down  | supply | -             | -             | "VSS"            ] ;| S   | -   | -      | -                                                                                                                   | -
+    [ VSS[7]         |  51 | Down  | supply | -             | -             | "VSS"            ] ;| S   | -   | -      | -                                                                                                                   | -
+    [ VSS[8]         |  83 | Down  | supply | -             | -             | "VSS"            ] ;| S   | -   | -      | -                                                                                                                   | -
+
+  make-box-symbol()
+  assign-landpattern(STM-LQFP144)
+
+  pcb-bundle SWD_SWDIO:
+    pin p
+  pcb-bundle SWD_SWDCLK:
+    pin p
+
+  supports SWD_SWDIO:
+    SWD_SWDIO.p => self.PA[13]
+  supports SWD_SWDCLK:
+    SWD_SWDCLK.p => self.PA[14]
+
+  supports swd() :
+    require swdio : SWD_SWDIO
+    require swdclk : SWD_SWDCLK
+    swd().swdio  => swdio.p
+    swd().swdclk => swdclk.p
+    property(swdio.p.stm32-signal-name) = "SYS_JTMS-SWDIO"
+    property(swdclk.p.stm32-signal-name) = "SYS_JTCK-SWCLK"
+    property(swdio.p.stm32-pin-mode)  = ocdb/st-microelectronics/generate-ioc/SerialWireDebugMode
+    property(swdclk.p.stm32-pin-mode) = ocdb/st-microelectronics/generate-ioc/SerialWireDebugMode
+
+  ; supports power:
+  ;   for i in 0 to 12 do :
+  ;     power.vdd => self.VDD[i]
+  ;   for i in 0 to 9 do :
+  ;     power.gnd => self.VSS[i]
+
+  supports reset:
+    reset.reset => self.NRST
+
+  val ADC12-LIST = [self.PA[0] self.PA[1] self.PA[2] self.PA[3] self.PA[4] self.PA[5] self.PA[6] self.PA[7] self.PB[0] self.PB[1] self.PC[0] self.PC[1] self.PC[2] self.PC[3] self.PC[4] self.PC[5]]
+  val ADC3-LIST = [self.PA[0] self.PA[1] self.PA[2] self.PA[3] self.PF[6] self.PF[7] self.PF[8] self.PF[9] self.PF[10] self.PF[3] self.PC[0] self.PC[1] self.PC[2] self.PC[3] self.PF[4] self.PF[5]]
+  val ADC3-UNIQUE-LIST = [4 5 6 7 8 9 14 15]
+
+  for (p in ADC12-LIST, i in 0 to false) do :
+    supports adc:
+      adc.adc => p
+      property(p.stm32-signal-name) = to-string("ADCx_IN%_" % [i])
+      property(p.stm32-ip-name) = ocdb/st-microelectronics/generate-ioc/ADC1IP
+      ;property(swdio.p.stm32-ip-name) = ocdb/st-microelectronics/generate-ioc/ADC2IP
+
+
+  for i in ADC3-UNIQUE-LIST do :
+    val p = ADC3-LIST[i]
+    supports adc:
+      adc.adc => p
+      property(p.stm32-signal-name) = to-string("ADC3_IN%_" % [i])
+      property(p.stm32-adc-mode) = to-string("IN%_" % [i])
+      property(p.stm32-ip-name) = ocdb/st-microelectronics/generate-ioc/ADC3IP
+
+  for i in 0 to 16 do :
```

Comment:
Sometimes ports like `PA` can be sparse (e.g. missing `PA[8]` and `PA[14]`) so you could use a code pattern like

```
for p in pins(self.PE) do :
  supports gpio :
    gpio.gpio => p
```

and save some difficulty.

Reply:
Ok, nice suggestion. I will incorporate it.

Reply:
Hmm.. I seem to have run into a problem with the pin property assignment:

  for p in pins(self.PA) do :
    supports gpio :
      gpio.gpio => p
      property(self.p.stm32-signal-name) = "GPIO"

complains as follows:
Uncaught Exception: Cannot access named field 'p'.
  in ocdb/st-microelectronics/STM32F407ZET6
    at ../open-components-database/components/st-microelectronics/STM32F407ZET6.stanza:253.15
  in ocdb/st-microelectronics/STM32F407ZET6
    at ../open-components-database/components/st-microelectronics/STM32F407ZET6.stanza:252.6
  in ocdb/st-microelectronics/STM32F407ZET6
    at ../open-components-database/components/st-microelectronics/STM32F407ZET6.stanza:250.25
  in ocdb/tests/test-stm32-mcu-props
    at test-stm32-mcu-props.stanza:40.7
  in ocdb/generator-utils/run-final-passes
    at ../open-components-database/utils/generator-utils.stanza:160.28
  in ocdb/tests/test-stm32-mcu-props
    at test-stm32-mcu-props.stanza:153.19

Reply:
Oh duh. Thanks for the fix.